### ISSUE TITLE
Elegance of dynamic imports in TypeScript

### ISSUE THREAD

**Issue by kachkaev** [zeit/next.js#4515]
# Bug report

## Describe the bug

This is a follow-up to https://github.com/zeit/next.js/issues/3389, which I cannot comment since the discussion is closed by the bot. The issue is to do with how dynamic imports work in TypeScript and it is still observed in Next 6.0.

## To Reproduce

The example is copy-pasted from a [comment](https://github.com/zeit/next.js/issues/3389#issuecomment-371978044) by @mohammedzamakhan:

```ts
import dynamic from "next/dynamic";

const Button = dynamic(import("../components/Button"));
```

The above code produces the following error:

```
Argument of type 'Promise<typeof "/components/Button" >' is not assignable to parameter of type 'Promise<ComponentType<{}>>'.
```

## Expected behavior

I expected `Button` to be a React component, just as when I use a non-dynamic import:

```ts
import Button from "../components/Button";
```

## Workarounds

1. Inspired by https://stackoverflow.com/a/49272129/1818285

    ```ts
    import unusedModule from "../components/Button";
    type ButtonType = typeof unusedModule;
    const Button = dynamic((import("../components/Button") as any) as Promise<ButtonType>);
    ```

2. Obtained through trial and error

    ```ts
    const Button = dynamic(import("../components/Button").then(module => module.default));
    ```

Both solutions are unintuitive, so I'm wondering if there is a more elegant way of correctly deriving component types during dynamic imports. Can it be the case that [@zeit/next-typescript](https://github.com/zeit/next-plugins/tree/master/packages/next-typescript) is suggesting a sub-optimal tsconfig (which I followed)? How to just make this notation work?

```ts
const Button = dynamic(import("../components/Button"));
``` 

## System information
 - OS: macOS
 - Version of Next.js: e.g. 6.0.3
 - @zeit/next-typescript: 1.0.1
 - tsconfig.json

```
{
  "compileOnSave": false,
  "compilerOptions": {
    "target": "esnext",
    "module": "esnext",
    "jsx": "preserve",
    "allowJs": true,
    "moduleResolution": "node",
    "allowSyntheticDefaultImports": true,
    "noUnusedLocals": true,
    "noUnusedParameters": true,
    "removeComments": false,
    "preserveConstEnums": true,
    "sourceMap": true,
    "skipLibCheck": true,
    "strictNullChecks": true,
    "baseUrl": ".",
    "lib": ["dom", "es2016"]
  },
  "include": ["src/**/*"]
}
```

**Comment by ex3ndr** [comment]
Isn't it you can just use plain import in typescript? I am using this for async module loading and it works well.

**Comment by kachkaev** [comment]
@ex3ndr could you please share an example of what you mean?

**Comment by ex3ndr** [comment]
@kachkaev 
Just like this: 
```
import("../components/Button").then((Button) => ...)
```

AFAIK, `dynamic` is just a wrapper for this promise and you can easily write it by yourself.

**Comment by alexxxPopa** [comment]
Does this work with custom loader component?

**Comment by timneutkens** [comment]
This will probably be fixed by the react-loadable rewrite

**Comment by timneutkens** [comment]
It's on canary only right now

**Comment by daraclare** [comment]
I am on canary and I am still seeing this error, is there something I'm missing?
```
"@types/next": "^7.0.5",
"@zeit/next-typescript": "^1.1.1",
 "next": "^7.0.2-canary.21",
```

```
const ImageGallery = dynamic(() => import('react-image-gallery'), {
  ssr: false,
  loading: () => null,
});
```

```
Argument of type '() => Promise<typeof import("/Users/myname/Code/myapp/node_modules/@types/react-image-gallery/index")>' is not assignable to parameter of type 'Promise<ComponentType<{}>>'.
  Property 'then' is missing in type '() => Promise<typeof import("/Users/myname/Code/myapp/node_modules/@types/react-image-gallery/index")>'.
```

Any update/advice would be much appreciated :)

**Comment by nvuhung** [comment]
Any update for this? I still have same issue :(

**Comment by DominikGuzei** [comment]
Having the same issue as @daraclare with Typescript and dynamic imports, none of the workarounds worked for me and i had to change the file extension to `.jsx` in the page to avoid typescript tripping over it (really not an ideal hack)

**Comment by Mosoc** [comment]
@daraclare @nvuhung @DominikGuzei 
I found a workaround just now.
```typescript
const DynamicComponent = dynamic(
(()=>import('MyComponents')) as any,
{ /* options */ },
)
```
I use it for non-SSR apollo-graphql query.
Please let me know whether it is helpful or not.

**Comment by DominikGuzei** [comment]
@Mosoc many thanks, this works for me 🎉

**Comment by resir014** [comment]
I know this is *really* late, but I've finally attempted to solve this issue once and for all. 🎉 

Please take a look here and give any feedback you have: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/33163

**Comment by LunaTK** [comment]
I also had same problem and solved it in a way of which make use of generics. 

First, you should define pops of your target component, Button in this case, if not exists.
```javascript
interface ButtonProps {
    onClick: () => any;
}
```
The interface above is just example, assuming that `onClick` is the only property used in `Button`.
Then, just specify this when calling dynamic import, and error will be gone.
```javascript
const Button = dynamic<ButtonProps>(import("../components/Button"));
```

**Comment by eric-burel** [comment]
Hi, 
Still hitting errors. Syntaxes that need to redefine the component interface are not satisfying, especially if you want to dynamically import 3rd party lib.

I hit:

```
82:14 Type '{ children: Element[]; className: string; center: any; zoom: number; }' is not assignable to type 'IntrinsicAttributes & { children?: ReactNode; }'.
  Property 'className' does not exist on type 'IntrinsicAttributes & { children?: ReactNode; }'.
    80 |       </ul>
    81 |       <div>
  > 82 |         <Map className={classes.mapRoot} center={position} zoom={13}>
```
for `dynamic(() => import("react-leaflet/lib/Map"), { ssr: false }`

**Comment by eric-burel** [comment]
Of course, `import { MapComponent } from "@types/react-leaflet";` to just import the file is forbidden...

**Comment by eric-burel** [comment]
Eventually I had too:
- Install `@types/react-leaflet` => to get the typings
- Copy paste  `node_modules/@types/react-leaflet/index.d.ts` in my code, as  `typings/react-leaflet.ts` => because you can't import `@types/react-leaflet`, it's forbidden by TypeScript. You are supposed to import `react-leaflet` to get typings, but here our initial issue is that it won't work server side.
- import types from there `import { MapProps } from "~/typings/react-leaflet"`
- `const Map = dynamic<MapProps>(() => import("react-leaflet/lib/Map"), { ssr: false})`

Now it works. It's quite circumvoluted but that somehow make sense.
I would however expect `dynamic` to be able to handle such packages. I'll probably open a new issue.

**Comment by ziishaned** [comment]
I fixed this issue by following:

```ts
const DynamicComponent = dynamic<HelloProps>(() =>
  import('../components/hello').then(mod => mod.Hello)
)
```

**Comment by eric-burel** [comment]
Yeah but how do you import HelloProps from a client only package. That's the issue here.

**Comment by liss-mouse** [comment]
I also faced the issue when trying to upgrade to the latest TS version. It works well with TS v3.5 but we want to upgrade TS. Now, we are blocked 😫

**Comment by eegli** [comment]
I found a workaround. This works for a **named export of an external library that requires props**. In my case, I tried to dynamically import `disqus-react`. But first of:

```tsx
const DynamicComponent = dynamic(
  () => import('disqus-react').then(mod => mod.DiscussionEmbed) as any,
  {
    ssr: false,
  },
)
```

This technically works but you lose all types for the library, and when you return it later on in your component, Typescript will complain again. **What you can do instead: Define a custom version of your component and re-export it as a default**.

```tsx
// custom-disqus.tsx

import { Blogpost } from '@/components/types';
import { DiscussionEmbed } from 'disqus-react';

type Props = Pick<Blogpost, 'id' | 'title' | 'slug'>;

const CustomDisqus: React.FC<Props> = ({ id, title, slug }) => {
  return (
    <DiscussionEmbed
      shortname='your-site'
      config={{
        url: `https://example.com/${slug}`,
        identifier: id,
        title: title,
      }}
    />
  );
};

export default CustomDisqus;
```
And then:

```tsx
// parent-component.tsx

const DynamicDisqus = dynamic(
  () => import('@/components/custom-disqus/custom-disqus.component'),
  {
    ssr: false,
  },
);

// ....

    return (
      <div>
        <DynamicDisqus id={post.id} title={post.title} slug={post.slug} />
      </div>
    );
```
This way, the compiler also knows what props your dynamically imported component expects. Works in Typescript >4.0.0.

**Comment by nnt25251325** [comment]
> I also had same problem and solved it in a way of which makes use of generics.
> 
> First, you should define pops of your target component, Button in this case, if not exists.
> 
> ```js
> interface ButtonProps {
>     onClick: () => any;
> }
> ```
> 
> The interface above is just example, assuming that `onClick` is the only property used in `Button`.
> Then, just specify this when calling dynamic import, and error will be gone.
> 
> ```js
> const Button = dynamic<ButtonProps>(import("../components/Button"));
> ```

Thanks
export default dynamic<any>(() => import(`@Components/Header/${SITE}`))

**Comment by jan-wilhelm** [comment]
This works perfectly for me, and still provides type safety:

```typescript
import type { Component as ComponentType } from "library";

const Component = dynamic(() =>
  import("library").then((lib) => lib.Component as any)
) as typeof ComponentType;


// ...

export default function () {
    return <Component>{/*...*/}</Component>
}
```

**Comment by abcd-ca** [comment]
Heads up, this worked for me but I needed `MyNoSSRComponent` to be the default export, not a named export.

```typescript
import {MyNoSSRComponentProps} from '../components/MyNoSSRComponent'

const MyNoSSRComponent = dynamic<MyNoSSRComponentProps>(
  () => import("../components/MyNoSSRComponent"), 
  { ssr: false}
)
```

**Comment by mydearxym** [comment]
> Heads up, this worked for me but I needed `MyNoSSRComponent` to be the default export, not a named export.
> 
> ```ts
> import {MyNoSSRComponentProps} from '../components/MyNoSSRComponent'
> 
> const MyNoSSRComponent = dynamic<MyNoSSRComponentProps>(
>   () => import("../components/MyNoSSRComponent"), 
>   { ssr: false}
> )
> ```

i think this is the best solution !

**Comment by Xetera** [comment]
While trying to figure out a solution to this I discovered a _reeally_ weird behavior.
This approach doesn't work and results in the problematic `React.ReactComponent<{}>` type that I imagine is the reason most people here are struggling with type inference
```ts
const DragDropContext = dynamic(
  () => import("react-beautiful-dnd").then((r) => r.DragDropContext),
  { ssr: false }
);
```
But for some weird reason this resolves into the correct `React.ReactComponent<DragDropContextProps>`
```ts
const DragDropContext = dynamic(
  () => {
    const promise = import("react-beautiful-dnd").then((r) => r.DragDropContext);
    return promise;
  },
  { ssr: false }
);
```
obviously the two should not have any differences which is really confusing to me. I can replicate this behavior with `react-beautiful-dnd` and custom components. I'm curious to see if anyone else can replicate this behavior.

**Comment by eric-burel** [comment]
Very interesting. Maybe related to something like this issue ? https://github.com/microsoft/TypeScript/issues/7220
TS might have different behavior when typing a return type and typing the "promise" assignment in your example.

**Comment by jesstelford** [comment]
This is what worked for me after some (_lots_) of Googling + a couple of creative attempts:

Specifically this is for named exports

```typescript
import React from 'react'
import dynamic from 'next/dynamic'

// 'import type' will get stripped by the TS compiler, so we wont pay
// any bundle size penalty for doing these kinds of imports (ie; we're NOT
// undoing the benefits of the 'dynamic()' call below)
import type { Button as ButtonType } from './button'

// This type represents the props + defaultProps of the above component
// See: https://stackoverflow.com/questions/43230765/typescript-react-access-component-property-types/55005902#comment107210124_55005902
type ButtonPropsType = JSX.LibraryManagedAttributes<
  typeof ButtonType,
  React.ComponentProps<typeof ButtonType>
>

// Lazy load this component to avoid pulling in huge JS bundles when not
// necessary
const Button = dynamic<ButtonPropsType>(() => import('./button').then(mod => mod.Button))
```

**Comment by thinkerelwin** [comment]
if you are using components without any props, you can use this instead:

const Button = dynamic<React.ReactNode>(() => import('./button').then(mod => mod.Button))

**Comment by aleksgribko** [comment]
My typescript was happy with this code

`
//Interface I import from the child component
export const ButtonWithNoSSR = dynamic<IButtonProps>(
  import('./Button.component') as any,
  {
    ssr: false,
  },
);`

**Comment by kmvan** [comment]
> if you are using components without any props, you can use this instead:
> 
> const Button = dynamic<React.ReactNode>(() => import('./button').then(mod => mod.Button))

to 
```
const Button = dynamic<React.ReactNode>(() => import('./button').then(({Button}) => Button))
```
Works and feels good.

**Comment by balazsorban44** [comment]
This issue has been automatically locked due to no recent activity. If you are running into a similar issue, please create a new issue with the steps to reproduce. Thank you.
